File: C/Queue/queue_using_linkedlist.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node *front=NULL, *rear=NULL;

void enqueue(int x) {
    struct Node* temp = (struct Node*)malloc(sizeof(struct Node));
    temp->data = x;
    temp->next = NULL;
    if(rear==NULL) { front=rear=temp; return; }
    rear->next = temp;
    rear = temp;
}

void dequeue() {
    if(front==NULL) { printf("Queue Underflow!\n"); return; }
    struct Node* temp = front;
    printf("Dequeued: %d\n", front->data);
    front = front->next;
    if(front==NULL) rear=NULL;
    free(temp);
}

void display() {
    if(front==NULL) { printf("Queue is empty\n"); return; }
    struct Node* temp = front;
    printf("Queue elements: ");
    while(temp!=NULL) { printf("%d ", temp->data); temp=temp->next; }
    printf("\n");
}

int main() {
    int choice, x;
    while(1) {
        printf("\n1.Enqueue 2.Dequeue 3.Display 4.Exit\nEnter choice: ");
        scanf("%d",&choice);
        switch(choice){
            case 1: printf("Enter element to enqueue: "); scanf("%d",&x); enqueue(x); break;
            case 2: dequeue(); break;
            case 3: display(); break;
            case 4: return 0;
            default: printf("Invalid choice\n");
        }
    }
}

File: python/Queue/queue_using_linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self):
        self.front = None
        self.rear = None

    def enqueue(self, data):
        new_node = Node(data)
        if not self.rear:
            self.front = self.rear = new_node
            return
        self.rear.next = new_node
        self.rear = new_node

    def dequeue(self):
        if not self.front:
            print("Queue Underflow!")
            return
        print("Dequeued:", self.front.data)
        self.front = self.front.next
        if not self.front:
            self.rear = None

    def display(self):
        if not self.front:
            print("Queue is empty")
            return
        curr = self.front
        elements = []
        while curr:
            elements.append(curr.data)
            curr = curr.next
        print("Queue elements:", elements)

# Driver
q = Queue()
while True:
    print("\n1.Enqueue 2.Dequeue 3.Display 4.Exit")
    choice = int(input("Enter choice: "))
    if choice == 1:
        element = int(input("Enter element to enqueue: "))
        q.enqueue(element)
    elif choice == 2:
        q.dequeue()
    elif choice == 3:
        q.display()
    elif choice == 4:
        break
    else:
        print("Invalid choice")
